Fix issues in accurateCastOrNull

aiven-sal opened the pull request:
`accurateCastOrNull` must not throw when casting an arbitrary string to Bool.

Close #38585

This PR also fixes an unrelated issue that caused `accurateCastOrNull` to behave like `accurateCastOrDefault` on error when casting to IPv4 or IPv6.   


### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* accurateCastOrNull will not throw when casting an arbitrary string to Bool and it'll return Null
* accurateCastOrNull will return Null instead of the all zeroes addresses, if an invalid string is casted to an IP
* When cast_keep_nullable=1 casting from a nullable empty string to IP will produce Null as it happens with all of the other types


<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

aiven-sal committed:
accurateCastOrNull must not throw when casting arbitrary string to Bool

Fix #38585

aiven-sal committed:
Return Null instead of default in accurateCastOrNull for invalid IPs

aiven-sal committed:
Fix unrelated format issues that clang-format complained about

aiven-sal committed:
Add tests

alexey-milovidov commented on src/Functions/FunctionsConversion.h:
```diff
@@ -1597,7 +1597,20 @@ struct ConvertImplGenericFromString
+                {
+                    serialization_from.deserializeWholeText(column_to, read_buffer, format_settings);
+                }
+                catch (Exception &)
```
This is wrong because there can be unrelated exceptions, such as memory errors.

alexey-milovidov replied:
Missing const.

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #54136 from aiven-sal/ClickHouse:aiven-sal/fix_38585 -->
This is an automated comment for commit bf47ac5f535e058eba7401d6379be1fdef201732 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54136/bf47ac5f535e058eba7401d6379be1fdef201732/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

aiven-sal committed:
Re-throw exceptions that are not related to bool parsing errors

alexey-milovidov reviewed (approved):
Yes, rethrowing non-specific exceptions is enough.

aiven-sal committed:
Fix style issue

aiven-sal committed:
Fix tests when cast_keep_nullable=1

Before this set of changes, when cast_keep_nullable=1,
CAST(toNullable(''), type),
accurateCast(toNullable(''), type),
accurateCastOrDefault(toNullable(''), type), and
accurateCastOrNull('', type)
produced NULL for all types
(at least for those where the empty string is invalid)
except IPv6 and IPv4.
Now they produce NULL also for IP types.

alexey-milovidov commented:
@aiven-sal Fuzzer has found an error, PTAL.
